### ISSUE TITLE
nan avg loss but works on pjreddie repo

### ISSUE THREAD

**Issue by simtrax** [AlexeyAB/darknet#2002]
Hi,

Sorry for posting yet another issue concerning nan values. But I can't seem to find an answer in previous issues.

I'm training plant recognition, small plants shot with a drone camera. I have cropped images in regions with many plants. Example image [here](https://imgur.com/a/C4I8hKk). Images vary in size and amount of plants, also zoom level varies.
I used yolo_mark to annotate images, I have rechecked all of them again, they seem fine. There is around 100 annotations of each plant (I know it is small, but this is just me testing stuff out, and it worked with pjreddies repo).
I have a bunch of negative images without annotations, just soil showing in the image.

yolo.cfg found [here](https://pastebin.com/J4jjGhnz), followed the guide in this repo exactly. I have two classes.
Output when training [here](https://pastebin.com/BiiQkQHX).

Following command is run when training: 
```
./darknet detector train data/obj.data cfg/yolo.cfg darknet53.conv.74
```

Hardware is a single 2080Ti.

The issue is that I get nan values for avg loss after around 5-7 steps when using this repo. 
Sometimes, I have got it to run longer by setting 256x256 and subdivisions higher (32 or 64).

So I almost gave up on this repo, I tried annotating new datasets, sometimes it work so I get avg loss 3-4 after 10k-20k steps. But it seems really unstable, most of my tries I get nan avg loss after less than 100 steps. Most of the times after 5-7 steps.

So I tried pjreddie repo, exactly the same dataset, with [this](https://pastebin.com/XBUCZhnC) cfg file.
And it works great, never any nan avg loss. And I trained to 0.3 avg loss in 5k steps.

So there's slight differences in cfg-files, but that can't be it. I tried higher resolutions in AlexeyAB repo, still nan avg loss values.

So what is going on here? To me this seems really strange.
Why do I get nan values in AlexeyAB repo and not in pjreddie repo, even though same dataset.

**Comment by AlexeyAB** [comment]
@simtrax 

What parameters do you use in the Makefile?

**Comment by simtrax** [comment]
@AlexeyAB 

You might be on to something there. So I have fiddled around a bit with different settings in the Makefile.
These settings creates the issue where I get nan values after just 6-7 steps.
```
GPU=1
CUDNN=1
CUDNN_HALF=1
OPENCV=1
AVX=0
OPENMP=0
LIBSO=0
```
If i set `CUDNN_HALF=0` it seems to work better. Right now 100 steps have passed and loss is getting lower. 
I will continue training tomorrow and run it longer.

I should be honest and say that I don't really know what the CUDNN_HALF parameter does. I read something about tensor cores. And my understanding is that the 2080Ti is equipped with tensor cores.

**Comment by AlexeyAB** [comment]
`CUDNN_HALF=1` should be used only for: Volta, Xavier, Turing or higher architecture: https://github.com/AlexeyAB/darknet/blob/21a4ec9390b61c0baa7ef72e72e59fa143daba4c/Makefile#L10


So you can use it for 2080Ti, but currently you should use it only after 2000 iterations of training.

I will fix it, so it will use fp32 for 1-2000 iterations, and fp16 for >2000 iteration automatically.

**Comment by simtrax** [comment]
Alright, thank you for the quick support!

**Comment by AlexeyAB** [comment]
@simtrax 

I fixed it. Not training goes using FP32 for the first 1000 iterations (if burin_in=1000 in cfg), and then continue training using FP16+FP32, using Tensor Cores.

Update your code, compile with `CUDNN_HALF=1` and train from the begining. Does it work?

**Comment by simtrax** [comment]
@AlexeyAB 

I ran a training session for 1000 iterations, everything was good. Down to 0.4 in loss.
Then Tensor Cores was used, at step 1013 i got nan values.
Pasting a bit of the output [here](https://pastebin.com/mPQakv2s).

It might be an issue with my dataset. I don't know.

**Comment by AlexeyAB** [comment]
@simtrax 

Try to set line
`if (state.index != 0 && state.net.cudnn_half && !l.xnor && (!state.train || iteration_num > state.net.burn_in*2))`

* here: https://github.com/AlexeyAB/darknet/blob/7bd6b2f7201baa6777d1f50652beb36578374e86/src/convolutional_kernels.cu#L298
* and here: https://github.com/AlexeyAB/darknet/blob/7bd6b2f7201baa6777d1f50652beb36578374e86/src/convolutional_kernels.cu#L478


If it doesn't help, try to set line:
`if (state.index != 0 && state.net.cudnn_half && !l.xnor && (!state.train || iteration_num > state.net.burn_in*3))`

If it will help, I will push this fix.

**Comment by simtrax** [comment]
@AlexeyAB 

It seems to work with this code:
`if (state.index != 0 && state.net.cudnn_half && !l.xnor && (!state.train || iteration_num > state.net.burn_in*2))`

What conclusion can I draw from this?

**Comment by AlexeyAB** [comment]
@simtrax I added this fix to my repo.

> What conclusion can I draw from this?

So your dataset is OK.

Just first 2000-3000 iterations is better to train by using Float-32 bit instead of Float-16 bit. Now it works automatically.

Do you train on Amazon EC2 or using your own server with connected monitor? 
In the last case you can train with `-map` flag and specified `valid=valid.txt` parameter in `obj.data` file to get mAP (mean average precision) - red line on the Loss-chart for each 4 Epochs: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

`./darknet detector train data/obj.data yolo-obj.cfg darknet53.conv.74 -map`

![68747470733a2f2f6873746f2e6f72672f776562742f79642f766c2f61672f7964766c616775746f66327a636e6a6f64737467726f656e3861632e6a706567](https://user-images.githubusercontent.com/4096485/49746959-69d96300-fcb3-11e8-95b6-0010e2f33237.jpg)

----

Later I will check what's value gets better accuracy (mAP) and reduces loss faster: `state.net.burn_in*2`, `state.net.burn_in*3`, ...

**Comment by simtrax** [comment]
Great! 
I'm training on my own computer with connected monitor. 

Thanks for all the help!
I'll close the issue.
